server/src/main/java/org/cloudfoundry/identity/uaa/authentication/BackwardsCompatibleTokenEndpointAuthenticationFilter.java

```diff
@@ -237,9 +238,15 @@ protected Authentication attemptTokenAuthentication(HttpServletRequest request,
                 log.debug("Attempting SAML authentication for token endpoint.");
                 try {
                     authResult = saml2BearerGrantAuthenticationConverter.convert(request);
+                } catch (AuthenticationException e) {
+                    String errorMessage = (e instanceof Saml2AuthenticationException saml2AuthenticationException) ?
+                            saml2AuthenticationException.getSaml2Error().getDescription() : e.getMessage();
+                    log.debug(errorMessage, e);
+                    throw new InsufficientAuthenticationException(errorMessage);
                 } catch (Exception e) {
-                    log.error("Error setting assertion in SAML filter", e);
-                    throw new InsufficientAuthenticationException("Error setting assertion in SAML filter");
+                    String errorMessage = "Error setting assertion in SAML filter";
```

Comment:
separated error situations...
In SAML bearer we authenticate a client and thus we can return more details about the problems, e.g. expired SAML assertion ...